test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,342 @@
+												Port: networking_v1.ServiceBackendPort{Number: 80},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		require.NoError(f.T(), f.Client.Create(context.TODO(), i))
+
+		By("ensuring it is routable")
+		checkRoutability(appHost)
+	})
+
+	AfterEach(func() {
+		By("cleaning up test artifacts")
+		f.DeleteNamespace(namespace)
+	})
+
+	Specify("applications remain routable after the upgrade", func() {
+		resources := updateContourDeploymentResources()
+
+		By("waiting for contour deployment to be updated")
+		require.Eventually(f.T(), func() bool {
+			require.NoError(f.T(), f.Client.Get(context.TODO(), client.ObjectKeyFromObject(resources.contourDeployment), resources.contourDeployment))
+			return resources.contourDeployment.Status.Replicas == *resources.contourDeployment.Spec.Replicas
+		}, time.Minute*1, time.Millisecond*50)
+
+		By("waiting for envoy daemonset to be updated")
+		require.Eventually(f.T(), func() bool {
+			require.NoError(f.T(), f.Client.Get(context.TODO(), client.ObjectKeyFromObject(resources.envoyDaemonSet), resources.envoyDaemonSet))
+			// This might work for now while we have only one worker node, but
+			// if we expand to more, we will have to rethink this.
+			return resources.envoyDaemonSet.Status.NumberReady > 0
+		}, time.Minute*3, time.Millisecond*50)
+		// TODO: when we deploy the from version in this test, ensure envoy
+		// shutdown time is configured to take less time.
+
+		By("ensuring app is still routable")
+		checkRoutability(appHost)
+	})
+})
+
+func ingressPathTypePtr(t networking_v1.PathType) *networking_v1.PathType {
+	return &t
+}
+
+func checkRoutability(host string) {
+	res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+		Host:      host,
+		Path:      "/echo",
+		Condition: e2e.HasStatusCode(200),
+	})
+	require.Truef(f.T(), ok, "expected 200 response code, got %d", res.StatusCode)
+}
+
+type contourDeploymentResources struct {
+	namespace                 *v1.Namespace
+	contourServiceAccount     *v1.ServiceAccount
+	envoyServiceAccount       *v1.ServiceAccount
+	contourConfigMap          *v1.ConfigMap
+	extensionServiceCRD       *apiextensions_v1.CustomResourceDefinition
+	httpProxyCRD              *apiextensions_v1.CustomResourceDefinition
+	tlsCertDelegationCRD      *apiextensions_v1.CustomResourceDefinition
+	certgenServiceAccount     *v1.ServiceAccount
+	contourRoleBinding        *rbac_v1.RoleBinding
+	certgenRole               *rbac_v1.Role
+	certgenJob                *batch_v1.Job
+	contourClusterRoleBinding *rbac_v1.ClusterRoleBinding
+	contourClusterRole        *rbac_v1.ClusterRole
+	contourService            *v1.Service
+	envoyService              *v1.Service
+	contourDeployment         *apps_v1.Deployment
+	envoyDaemonSet            *apps_v1.DaemonSet
+}
+
+// Unmarshals resources from rendered Contour manifest in order and updates
+// each.
+func updateContourDeploymentResources() contourDeploymentResources {
```

Comment:
This is really neat, seems like it will need to be updated if we add any bits to the example YAML though? If so, Not a dealbreaker, since the job will fail then, but maybe worth a comment?